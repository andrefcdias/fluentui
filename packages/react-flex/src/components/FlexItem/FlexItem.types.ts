import * as React from 'react';
import { ComponentProps, ComponentState } from '@fluentui/react-utilities';
import { GlobalsNumber, AlignSelfProperty, FlexBasisProperty } from 'csstype';

/**
 * FlexItem component properties
 * {@docCategory FlexItem }
 */
export interface FlexItemProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * Self alignment in flex layout.
   * Wrapper of the CSS `align-self`.
   */
  align?: AlignSelfProperty;

  /**
   * Grow factor.
   * Wrapper of the CSS `flex-grow`.
   */
  grow?: GlobalsNumber;

  /**
   * Shrink factor.
   * Wrapper of the CSS `flex-shrink`.
   */
  shrink?: GlobalsNumber;

  /**
   * Initial main size of a the item.
   * Wrapper of the CSS `flex-basis`.
   */
  basis?: FlexBasisProperty<string | number>;

  /**
   * Order to lay out itself in the flex container.
   * Wrapper of the CSS `order`.
   */
  order?: GlobalsNumber;
}

/**
 * FlexItem component state
 * {@docCategory FlexItem }
 */
export type FlexItemState = ComponentState<React.Ref<HTMLElement>, FlexItemProps>;
