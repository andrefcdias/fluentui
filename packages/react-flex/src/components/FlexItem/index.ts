export * from './FlexItem';
export * from './FlexItem.types';
export * from './renderFlexItem';
export * from './useFlexItem';
export * from './useFlexItemStyles';
