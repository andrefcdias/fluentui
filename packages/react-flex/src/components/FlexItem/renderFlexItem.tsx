import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { FlexItemState } from './FlexItem.types';

/**
 * Render the final JSX of FlexItem
 * {@docCategory FlexItem }
 */
export const renderFlexItem = (state: FlexItemState) => {
  const { slots, slotProps } = getSlots(state);

  return <slots.root {...slotProps.root}>{state.children}</slots.root>;
};
