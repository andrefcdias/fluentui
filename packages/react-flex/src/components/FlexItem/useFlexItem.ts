import * as React from 'react';
import { makeMergeProps, useMergedRefs } from '@fluentui/react-utilities';
import { FlexItemProps, FlexItemState } from './FlexItem.types';

const mergeProps = makeMergeProps<FlexItemState>();

/**
 * Create the state required to render FlexItem.
 *
 * The returned state can be modified with hooks such as useFlexItemStyles,
 * before being passed to renderFlexItem.
 *
 * @param props - props from this instance of FlexItem
 * @param ref - reference to root HTMLElement of FlexItem
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory FlexItem }
 */
export const useFlexItem = (
  props: FlexItemProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: FlexItemProps,
): FlexItemState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
    },
    defaultProps,
    props,
  );

  return state;
};
