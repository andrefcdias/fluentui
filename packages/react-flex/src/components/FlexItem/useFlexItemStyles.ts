import { mergeClasses, makeStyles } from '@fluentui/react-make-styles';
import { FlexItemState } from './FlexItem.types';

const useStyles = makeStyles({
  root: {},
});

/**
 * Apply styling to the FlexItem slots based on the state
 * {@docCategory FlexItem }
 */
export const useFlexItemStyles = (state: FlexItemState): FlexItemState => {
  const styles = useStyles();
  state.className = mergeClasses(styles.root, state.className);

  state.style = {
    ...state.style,
    alignSelf: state.align ?? 'auto',
    flexGrow: state.grow ?? 0,
    flexShrink: state.shrink ?? 1,
    flexBasis: state.basis ?? 'auto',
    order: state.order ?? 0,
  };

  return state;
};
