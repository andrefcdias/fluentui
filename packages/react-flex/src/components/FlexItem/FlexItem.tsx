import * as React from 'react';
import { useFlexItem } from './useFlexItem';
import { FlexItemProps } from './FlexItem.types';
import { renderFlexItem } from './renderFlexItem';
import { useFlexItemStyles } from './useFlexItemStyles';

/**
 * TODO: Component docs
 * {@docCategory FlexItem }
 */
export const FlexItem = React.forwardRef<HTMLElement, FlexItemProps>((props, ref) => {
  const state = useFlexItem(props, ref);

  useFlexItemStyles(state);
  return renderFlexItem(state);
});

FlexItem.displayName = 'FlexItem';
