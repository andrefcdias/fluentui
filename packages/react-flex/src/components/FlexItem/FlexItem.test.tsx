import * as React from 'react';
import { FlexItem } from './FlexItem';
import { render } from '@testing-library/react';
import { isConformant } from '../../common/isConformant';

describe('FlexItem', () => {
  isConformant({
    Component: FlexItem,
    displayName: 'FlexItem',
  });

  it('does not change node structure', () => {
    const { container } = render(<FlexItem>Item</FlexItem>);

    expect(container).toMatchSnapshot();
  });

  it('renders a default state', () => {
    const utils = render(<FlexItem>Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      order: 0,
      alignSelf: 'auto',
      flexBasis: 'auto',
      flexGrow: 0,
      flexShrink: 1,
    });
  });

  it('sets alignment', () => {
    const utils = render(<FlexItem align="flex-start">Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      alignSelf: 'flex-start',
    });
  });

  it('sets grow', () => {
    const utils = render(<FlexItem grow={1}>Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      flexGrow: 1,
    });
  });

  it('sets shrink', () => {
    const utils = render(<FlexItem shrink={0}>Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      flexShrink: 0,
    });
  });

  it('sets basis', () => {
    const utils = render(<FlexItem basis="10px">Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      flexBasis: '10px',
    });
  });

  it('sets order', () => {
    const utils = render(<FlexItem order={2}>Item</FlexItem>);

    const item = utils.getByText('Item');
    expect(item).toHaveStyle({
      order: 2,
    });
  });
});
