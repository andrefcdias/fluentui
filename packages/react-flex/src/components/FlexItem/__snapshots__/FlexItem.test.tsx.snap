// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexItem does not change node structure 1`] = `
<div>
  <div
    class=""
    style="align-self: auto; flex-grow: 0; flex-shrink: 1; flex-basis: auto; order: 0;"
  >
    Item
  </div>
</div>
`;
